aiida_cp2k/utils.py

```diff
@@ -74,23 +89,47 @@ def _render_section(output, params, indent=0):
                   &END KIND
         """
 
+        if six.PY2:
+            from collections import Mapping, Sequence
+        else:
+            from collections.abc import Mapping, Sequence
+
         for key, val in sorted(params.items()):
+            # the `_` is reserved for section params and evaluated in the prior call
+            if key == "_":
+                continue
+
+            # keys are not case-insensitive, ensure that they follow the current scheme
             if key.upper() != key:
-                raise ValueError("keyword '%s' not upper case" % key)
-            if key.startswith("@") or key.startswith("$"):
-                raise ValueError("CP2K preprocessor not supported")
-            if isinstance(val, dict):
-                line = "%s&%s" % (" " * indent, key)
+                raise ValueError("keyword '{key}' not upper case".format(key=key))
+
+            if key.startswith(("@", "$")):
+                raise ValueError("CP2K preprocessor directives not supported")
+
+            ispace = " " * indent
+
+            if isinstance(val, Mapping):
+                line = "{ispace}&{key}".format(ispace=ispace, key=key)
                 if "_" in val:  # if there is a section parameter, add it
-                    line += " %s" % val["_"]
+                    line += " {}".format(val["_"])
+
                 output.append(line)
-                Cp2kInput._render_section(output, val, indent + 3)
-                output.append("%s&END %s" % (" " * indent, key))
-            elif isinstance(val, list):
+                Cp2kInput._render_section(output, val, indent + indent_width)
```

Comment:
> A couple of questions/comments for [dcc9fce](https://github.com/aiidateam/aiida-cp2k/commit/dcc9fce12ee4d951d2eaeca8d04135ea8eb49b9f):
> 
> 1. Again, I would suggest to squash it into [0acf6db](https://github.com/aiidateam/aiida-cp2k/commit/0acf6db0deee6444c812ffdb313c9d634fdac04b), where there '_'-related bug was introduced (and the `render()` function has already become idempotent)

I didn't introduce a bug in 0acf6db. In dcc9fce I introduced a test which would exhibit the bug already present before 0acf6db plus the fix. Squashing that would hide the bugfixing in the refactoring.

> 2. I think the commit message should be modified, telling more about python2/3 compatibility introduced here.

There is no real python2/3 compatibility changes at this point. The only change "hidden" here is that instead of assuming the nested dictionary to consist of `dict` and `list`, it can now also consist of things implementing a `dict` or a `list` (like an `OrderedDict` or a `tuple`).

> 3. Just wondering: why do you always use such a long statements:
>    `"{ispace}{key} {val}".format(ispace=ispace, key=key, val=val)`?
>    why not just:
>    `"{}{} {}".format(ispace, key, val)
>    I think the strings are so short, that this cannot cause any confusions. Moreover, code becomes shorter.

This might be worth an additional comment, since it is a preparation for Python 3 f-strings. Once Python 2 compatibility can be dropped, those strings can simply be changed from  `"{ispace}{key} {val}".format(ispace=ispace, key=key, val=val)` to  `f"{ispace}{key} {val}"`, making the conversion much less error prone.

> 4. To me the clearest and simplest way to introduce idempotence is to use 2 times `deepcopy`: one time in `__init__` and another time in `render()`

If possible with a small algorithmic choice I usually prefer zero-copy where possible for efficiency reason.

Reply:
> Why do you want to let users modify `indent_width` value, does it ever needed? In addition this can potentially cause some tests to fail.

I usually prefer to move magic constants as far up in the logic as possible. While it makes sense that a function rendering the dictionary has a default, the caller of said function should eventually be in control of that constant.
Since `_render_section` is a protected function (`_` prefix) and not tested separately there is no risk of tests suddenly failing.